File: apex_dashboard/apex_dashboard/page/equity_profit_dashboard/equity_profit_dashboard.py
from __future__ import annotations

from datetime import date
from typing import Dict, List, Optional

import frappe
from frappe import _
from frappe.utils import add_months, format_date, get_first_day, getdate, nowdate

from apex_dashboard.dashboard import utils as dashboard_utils


@frappe.whitelist()
def get_context(context):
	context.no_cache = 1
	context.page_title = _("Equity & Profit Dashboard")
	return context


@frappe.whitelist()
def get_dashboard_data(company: Optional[str] = None, posting_date: Optional[str] = None):
	try:
		posting_date = posting_date or nowdate()
		posting_date = str(getdate(posting_date))

		data = build_equity_snapshot(company=company, posting_date=posting_date)

		return {
			"success": True,
			"filters": {
				"company": company,
				"posting_date": posting_date,
			},
			"data": data,
		}
	except Exception as exc:
		frappe.log_error(frappe.get_traceback(), "Equity & Profit Dashboard - get_dashboard_data")
		return {
			"success": False,
			"error": _(str(exc)),
		}


def build_equity_snapshot(company: Optional[str], posting_date: str) -> Dict[str, object]:
	mapping = dashboard_utils.get_accounts_for_section("equity_profit_dashboard")

	capital_balances = dashboard_utils.get_balances_for_accounts(mapping.get("capital", []), company, posting_date)
	owner_balances = dashboard_utils.get_balances_for_accounts(mapping.get("owner_equity", []), company, posting_date)
	retained_balances = dashboard_utils.get_balances_for_accounts(
		mapping.get("retained_earnings", []), company, posting_date
	)
	net_income_balances = dashboard_utils.get_balances_for_accounts(mapping.get("net_income", []), company, posting_date)

	capital_totals = _totals(capital_balances)
	owner_totals = _totals(owner_balances)
	retained_totals = _totals(retained_balances)
	net_income_totals = _totals(net_income_balances)

	kpis = [
		{
			"key": "capital",
			"label": _("رأس المال"),
			"totals": capital_totals,
			"indicator": "info",
		},
		{
			"key": "owner_equity",
			"label": _("حقوق الشركاء"),
			"totals": owner_totals,
			"indicator": "info",
		},
		{
			"key": "retained",
			"label": _("الأرباح المحتجزة"),
			"totals": retained_totals,
			"indicator": "positive" if retained_totals["base"] >= 0 else "warning",
		},
		{
			"key": "net_income",
			"label": _("صافي الدخل"),
			"totals": net_income_totals,
			"indicator": "positive" if net_income_totals["base"] >= 0 else "danger",
		},
	]

	chart = _build_income_trend(mapping.get("net_income", []), company, posting_date)

	return {
		"kpis": kpis,
		"capital": _format_group(_("رأس المال"), capital_balances, capital_totals),
		"owner_equity": _format_group(_("حقوق الشركاء"), owner_balances, owner_totals),
		"retained_earnings": _format_group(_("الأرباح المحتجزة"), retained_balances, retained_totals),
		"net_income": _format_group(_("صافي الدخل"), net_income_balances, net_income_totals),
		"chart": chart,
	}


def _totals(balances: List[dashboard_utils.AccountBalance]) -> Dict[str, object]:
	return {
		"by_currency": dashboard_utils.summarize_balances_by_currency(balances),
		"base": dashboard_utils.summarize_base_total(balances),
	}


def _format_group(label: str, balances: List[dashboard_utils.AccountBalance], totals: Dict[str, object]) -> Dict[str, object]:
	return {
		"label": label,
		"totals": totals,
		"balances": [balance.__dict__ for balance in balances],
	}


def _build_income_trend(accounts: List[str], company: Optional[str], posting_date: str) -> Dict[str, object]:
	if not accounts:
		return {"labels": [], "values": [], "currency": dashboard_utils.get_company_currency(company)}

	end_date = getdate(posting_date)
	labels: List[str] = []
	values: List[float] = []

	for idx in range(5, -1, -1):
		month_end = add_months(end_date, -idx)
		month_start = get_first_day(month_end)

		amount = _get_period_income(accounts, company, month_start, month_end)
		labels.append(format_date(month_end, "MMM yyyy"))
		values.append(amount)

	return {
		"labels": labels,
		"values": values,
		"currency": dashboard_utils.get_company_currency(company),
	}


def _get_period_income(accounts: List[str], company: Optional[str], from_date: date, to_date: date) -> float:
	conditions = [
		"gl.account in %(accounts)s",
		"gl.docstatus = 1",
		"gl.posting_date BETWEEN %(from_date)s AND %(to_date)s",
	]
	values = {
		"accounts": tuple(accounts),
		"from_date": from_date,
		"to_date": to_date,
	}
	if company:
		conditions.append("gl.company = %(company)s")
		values["company"] = company

	query = f"""
		SELECT
			COALESCE(SUM(gl.credit - gl.debit), 0) AS amount
		FROM `tabGL Entry` gl
		WHERE {' AND '.join(conditions)}
	"""

	result = frappe.db.sql(query, values=values, as_dict=True)
	return result[0].amount if result else 0.0

File: apex_dashboard/apex_dashboard/page/equity_profit_dashboard/equity_profit_dashboard.js
frappe.pages["equity-profit-dashboard"].on_page_load = function (wrapper) {
	frappe.require(
		[
			"assets/apex_dashboard/js/dashboard_common.js",
			"assets/apex_dashboard/css/dashboard_common.css",
		],
		() => init_equity_dashboard(wrapper)
	);
};

function init_equity_dashboard(wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __("Equity & Profit Dashboard"),
		single_column: true,
	});

	const state = {
		filters: {
			company: frappe.defaults.get_user_default("company"),
			posting_date: frappe.datetime.get_today(),
		},
		data: null,
		chart: null,
	};

	const layout = build_equity_layout(page);
	setup_equity_filters(page, state, refresh);

	function refresh() {
		page.set_indicator(__("جارِ التحميل..."), "blue");
		frappe
			.call("apex_dashboard.apex_dashboard.page.equity_profit_dashboard.equity_profit_dashboard.get_dashboard_data", {
				company: state.filters.company,
				posting_date: state.filters.posting_date,
			})
			.then((response) => {
				const payload = response.message || {};
				if (!payload.success) {
					throw new Error(payload.error || __("فشل تحميل بيانات حقوق الملكية."));
				}
				state.data = payload.data || {};
				state.filters = Object.assign({}, state.filters, payload.filters || {});
				render_equity_dashboard(layout, state);
				page.set_indicator(__("جاهز"), "green");
			})
			.catch((error) => {
				console.error(error);
				page.set_indicator(__("خطأ"), "red");
				frappe.msgprint({
					title: __("خطأ"),
					message: error.message || __("تعذر تحميل لوحة حقوق الملكية."),
					indicator: "red",
				});
			});
	}

	refresh();
}

function build_equity_layout(page) {
	const $container = $(page.body);
	$container.empty();

	const $wrapper = $(`
		<div class="gt-dashboard-wrapper">
			<section class="gt-dashboard-section">
				<div class="gt-dashboard-cards" data-role="kpis"></div>
			</section>
			<section class="gt-dashboard-section" data-role="chart">
				<div class="gt-dashboard-subcard">
					<h5>${__("اتجاه صافي الدخل (آخر 6 أشهر)")}</h5>
					<div id="equity-profit-chart" style="min-height: 260px;"></div>
				</div>
			</section>
			<section class="gt-dashboard-section" data-role="sections"></section>
		</div>
	`);

	$container.append($wrapper);

	return {
		kpis: $wrapper.find('[data-role="kpis"]')[0],
		chart: $wrapper.find("#equity-profit-chart")[0],
		sections: $wrapper.find('[data-role="sections"]')[0],
	};
}

function setup_equity_filters(page, state, refresh) {
	const company_field = page.add_field({
		fieldtype: "Link",
		fieldname: "company",
		label: __("شركة"),
		options: "Company",
		default: state.filters.company,
		change() {
			state.filters.company = company_field.get_value();
			refresh();
		},
	});

	const date_field = page.add_field({
		fieldtype: "Date",
		fieldname: "posting_date",
		label: __("تاريخ"),
		default: state.filters.posting_date,
		change() {
			state.filters.posting_date = date_field.get_value();
			refresh();
		},
	});

	page.add_action_icon("refresh", () => refresh());
}

function render_equity_dashboard(layout, state) {
	render_equity_kpis(layout.kpis, state.data.kpis || []);
	render_equity_sections(layout.sections, state.data);
	render_equity_chart(layout.chart, state);
}

function render_equity_kpis(container, kpis) {
	if (!container) return;
	const cards = kpis.map((kpi) => {
		const totals = kpi.totals || {};
		const baseValue = totals.base || 0;
		const byCurrency = totals.by_currency || {};
		const subtitle = Object.keys(byCurrency)
			.map((currency) => DashboardCommon.formatCurrency(byCurrency[currency], currency))
			.join(" • ");

		return {
			title: kpi.label || "",
			value: DashboardCommon.formatCurrency(
				baseValue,
				frappe.defaults.get_default("currency") || frappe.sys_defaults.currency || "EGP"
			),
			subtitle,
			indicator: kpi.indicator || "info",
		};
	});

	DashboardCommon.renderCards(container, cards);
}

function render_equity_sections(container, data) {
	if (!container) return;
	container.innerHTML = "";

	const sections = [
		data.capital,
		data.owner_equity,
		data.retained_earnings,
		data.net_income,
	].filter(Boolean);

	if (!sections.length) {
		container.innerHTML = `<div class="gt-dashboard-empty">${__("لا توجد بيانات تفصيلية")}</div>`;
		return;
	}

	sections.forEach((section) => {
		const card = document.createElement("div");
		card.className = "gt-dashboard-subcard";

		const header = document.createElement("div");
		header.className = "gt-dashboard-section__header";
		const title = document.createElement("h4");
		title.className = "gt-dashboard-section__title";
		title.innerText = section.label || "";
		header.appendChild(title);
		header.appendChild(DashboardCommon.buildTotalsRow(section.totals || {}));
		card.appendChild(header);

		const tableContainer = document.createElement("div");
		card.appendChild(tableContainer);
		DashboardCommon.renderGroupTable(tableContainer, section.balances || []);

		container.appendChild(card);
	});
}

function render_equity_chart(container, state) {
	if (!container || !state.data) return;

	const chartData = state.data.chart || {};
	const labels = chartData.labels || [];
	const values = chartData.values || [];

	if (!labels.length) {
		container.innerHTML = `<div class="gt-dashboard-empty">${__("لا توجد بيانات للمخطط")}</div>`;
		return;
	}

	if (state.chart) {
		state.chart.destroy();
		state.chart = null;
	}

	state.chart = new frappe.Chart(container, {
		data: {
			labels,
			datasets: [
				{
					values,
				},
			],
		},
		type: "line",
		height: 260,
		colors: ["#10b981"],
		axisOptions: {
			xAxisMode: "tick",
			yAxisMode: "tick",
			xIsSeries: true,
		},
		tooltipOptions: {
			formatTooltipY: (value) =>
				DashboardCommon.formatCurrency(value, chartData.currency || frappe.sys_defaults.currency || "EGP"),
		},
	});
}

